specs/beacon-chain.md

```diff
@@ -304,61 +304,108 @@ def get_active_validator_indices(validators):
 Now, a function that shuffles this list:
 
 ```python
-def shuffle(lst, seed):
+def shuffle(values: List[Any],
+            seed: Hash32) -> List[Any]:
+    """
+    Returns the shuffled ``values`` with seed as entropy.
+    """
+    values_count = len(values)
+
     # entropy is consumed in 3 byte chunks
     # rand_max is defined to remove the modulo bias from this entropy source
-    rand_max = 2**24
-    assert len(lst) <= rand_max
+    rand_max = 2 ** 24
+    assert values_count <= rand_max:
 
-    o = [x for x in lst]
+    output = [x for x in values]
     source = seed
-    i = 0
-    while i < len(lst):
-        source = hash(source)
-        for pos in range(0, 30, 3):
-            m = int.from_bytes(source[pos:pos+3], 'big')
-            remaining = len(lst) - i
+    index = 0
+    while index < values_count:
+        # Re-hash the source
+        source = blake(source)
+        for position in range(0, 30, 3):  # gets indices 3 bytes at a time
+            # Select a 3-byte sampled int
+            sample_from_source = int.from_bytes(source[position:position + 3], 'big')
+            # `remaining` is the size of remaining indices of this round
+            remaining = values_count - index
             if remaining == 0:
```

Comment:
Optimally this should be `remaining == 1`. If there's only 1 element left the only option is to swap it with itself. It doesn't change the result of the shuffle, it's simply an unnecessary iteration.

Reply:
And since this is a spec, most part of implementers will just port this algorithm to their languages without thinking much on how it exactly works. Would be great to have it optimal just in case.